taichi/ui/backends/vulkan/renderables/mesh.cpp

```diff
@@ -88,21 +88,8 @@ void Mesh::update_data(const MeshInfo &info, const Scene &scene) {
       attr_dev_ptr =
           get_device_ptr(prog, info.mesh_attribute_info.mesh_attribute.snode);
     }
-    Device::MemcpyCapability memcpy_cap = Device::check_memcpy_capability(
```

Comment:
Why don't we check capability anymore?

Reply:
Because if we add the function `memcpy_cpu_to_vulkan`, then the `memcpy_direct` includes all kinds of situations as below:
1. copy from cpu data to vulkan staging buffer
2. copy from cuda data to vulkan staging buffer
3. copy from vulkan data to vulkan buffer
4. others will report `ti_not_implemented`

Reply:
what about situations like `opengl` or `metal`, which should go from opengl->cpu->vulkan before we implement the interop

Reply:
Yes, I think that's the 4th situation, GGUI doesn't handle opengl and metal for now, and we report an error info. Maybe in future, we should implement that. By the way, I think `check_memcpy_capability` is superfluous, because all other copy functions (like `memcpy via staging` and `memcpy_direct`) have the same checkings of what the `src` and `dst` belong to. Why should we check that twice?

Reply:
Ok, I forgot that, that's right, but for now, we don't support `opengl` and `metal` even using memcpy_via_staging, which is a future work I think? 
By the way I think `check_memcpy_capability` is really confused, because we have to check the `src` and `dst` device twice. Why we should check with dynamic_cast instead of tagging the device backend?

Reply:
How can we tag the backend?

Reply:
Maybe add an enum class type in child class I think?

Reply:
I think dynamic cast is not really slower than that?